Decorators 101 Guide (#302)
astronomer/airflow-guides: guides/airflow-decorators.md

In this guide, we'll cover the benefit of decorators, the decorators available in Airflow, and decorators provided in Astronomer's open source `astro` library. We'll also show examples throughout and address common questions about when to use decorators and how to combine them with traditional operators in a DAG.

## When and Why To Use Decorators

The purpose of decorators in Airflow is to simplify the DAG authoring experience by eliminating the boilerplate code required by traditional operators. The result can be cleaner DAG files that are more concise and easier to read. Currently, decorators can be used for Python and SQL functions.

In general, whether to use decorators is a matter of developer preference and style. Generally, a decorator and the corresponding traditional operator will have the same functionality. One exception to this is the `astro` library of decorators (more on these below), which do not have equivalent traditional operators. You can also easily mix decorators and traditional operators within your DAG if your use case requires that.

## How to Use Airflow Decorators

> Airflow decorators were introduced as part of the TaskFlow API, which also handles passing data between tasks using XCom and inferring task dependencies automatically. To learn more about the TaskFlow API, check out [Astronomer's webinar](https://www.astronomer.io/events/webinars/taskflow-api-airflow-2.0) or the Apache Airflow [Taskflow API tutorial](https://airflow.apache.org/docs/apache-airflow/stable/tutorial_taskflow_api.html#tutorial-on-the-taskflow-api). 

Reviewer: ```suggestion
Airflow decorators were introduced as part of the TaskFlow API, which also handles passing data between tasks using XCom and inferring task dependencies automatically. To learn more about the TaskFlow API, check out [Astronomer's webinar](https://www.astronomer.io/events/webinars/taskflow-api-airflow-2.0) or the Apache Airflow [TaskFlow API tutorial](https://airflow.apache.org/docs/apache-airflow/stable/tutorial_taskflow_api.html#tutorial-on-the-taskflow-api). 
```
To be consistent with the spelling of "TaskFlow". I agree on using this particular spelling too.

- You can decorate a function that is imported from another file with something like the following:

    ```python
    from include.my_file import my_function

    @task
    def taskflow_func():
        my_function()
    ```
    
    This is recommended in cases where you have lengthy Python functions since it will make your DAG file easier to read. 

### Mixing Decorators with Traditional Operators

If you have a DAG that uses `PythonOperator` and other operators that don't have decorators, you can easily combine decorated functions and traditional operators in the same DAG. For example, we can add an `EmailOperator` to the previous example by updating our code to the following:

```python
from airflow.decorators import dag, task
from airflow.utils.dates import days_ago
from airflow.operators.email_operator import EmailOperator

from typing import Dict
import requests
import logging

API = "https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=usd&include_market_cap=true&include_24hr_vol=true&include_24hr_change=true&include_last_updated_at=true"

default_args = {
    'start_date': days_ago(1),

> }

Reviewer: Let's get rid of declaring `start_date` in `default_args` and declare right in the DAG object. And also use a static start date. I know this is the "before-and-after" section and not the crux but IMO we should at least show ideal authoring even if it's not the most up-to-date authoring style. WDYT?

Author: Dang, I thought I had fixed that already. Yes, will definitely update - this isn't supposed to be a "bad practices" example, just one without TaskFlow

The library contains SQL and dataframe decorators that greatly simplify your DAG code and allow you to directly define tasks without boilerplate operator code. It also allows you to transition seamlessly between SQL and Python for transformations without having to explicitly pass data between tasks or convert the results of queries to dataframes and vice versa. For a full description of functionality, check out the [repo Readme](https://github.com/astro-projects/astro).

To use the `astro` library, you need to install the `astro-projects` package in your Airflow environment and enable pickling (`AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True`). 

To show `astro` in action, we'll use a simple ETL example. We have data on adoptions from two different animal shelters that we need to aggregate, clean, transform, and append to a reporting table. Some of these tasks are better suited to SQL, and some to Python, but we can easily combine both using `astro` functions. The DAG looks like this:

```python
from airflow.decorators import dag

> from astro import sql as aql

Reviewer: I know this has been the convention in the README examples for the `astro` repo but I feel it's actually more readable to keep it as `from astro import sql` and then use `@sql.transform`. Or even importing this directly via `from astro.sql import transform` and changing the decoration. WDYT?

Author: My guess is the decision to use `aql` as an alias is because `sql` might also be used for other things and that could get confusing. But I agree with you on readability, changed it to `from astro.sql import transform, append` and updated the decoration.

from astro.sql.table import Table

> from astro import dataframe as df

Reviewer: Might be more readable later on to use `from astro import dataframe`?

Author: Yeah, I do like this better. Using `df` as an alias for dataframe is pretty standard, but it does get a bit confusing in the example when you also have `pandas` dataframes that are being referenced as `df`

from datetime import datetime, timedelta
import pandas as pd

# Start by selecting data from two source tables in Snowflake

> @aql.transform()

Reviewer: ```suggestion
@aql.transform
```
Should be fine to not use the parentheses as if you were calling the function.

def combine_data(center_1: Table, center_2: Table):

>     return """SELECT * FROM {center_1}
>     UNION SELECT * FROM {center_2}"""

Reviewer: ```suggestion
    return """
    SELECT * FROM {center_1}
    UNION
    SELECT * FROM {center_2}
    """
```
This feels a little more readable. WDYT?

Author: Yep, agreed

# Clean data using SQL

> @aql.transform()

Reviewer: Same parentheses comment here. (Unless for some reason this breaks functionally.)

Author: It's fine to leave them off. I think I had initially put them on to show adding `retries` or some other task-level param in this DAG, but that hasn't been implemented yet for `astro`. Agree it's cleaner to leave them off if we aren't passing anything in.

def clean_data(input_table: Table):

>     return '''SELECT * 
>     FROM {input_table} WHERE TYPE NOT LIKE 'Guinea Pig'
>     '''

Reviewer: ```suggestion
    return '''
    SELECT * 
    FROM {input_table} WHERE TYPE NOT LIKE 'Guinea Pig'
    '''
```
Similar styling here too. No strong opinions though.

default_args = {
    'start_date': days_ago(1),
}

@dag(schedule_interval='@daily', default_args=default_args, catchup=False)
def taskflow():

    @task(task_id='extract', retries=2)
    def extract_bitcoin_price() -> Dict[str, float]:
        return requests.get(API).json()['bitcoin']

    @task(multiple_outputs=True)
    def process_data(response: Dict[str, float]) -> Dict[str, float]:
        logging.info(response)
        return {'usd': response['usd'], 'change': response['usd_24h_change']}

    @task
    def store_data(data: Dict[str, float]):
        logging.info(f"Store: {data['usd']} with change {data['change']}")

    store_data(process_data(extract_bitcoin_price()))

dag = taskflow()
```

The resulting DAG has much less code and is easier to read. Notice that it also doesn't require us to use `ti.xcom_pull` and `ti.xcom_push` to explicitly pass data between our tasks. This is all handled by the TaskFlow API when we define our task dependencies with `store_data(process_data(extract_bitcoin_price()))`. 

Here are some other things to keep in mind when using decorators:

> - For any decorated object in your DAG file, you must call them for Airflow to register the task or DAG (e.g. `dag = taskflow()`).

Reviewer 2: ```suggestion
- For any decorated object in your DAG file, you must call them so that Airflow can register the task or DAG (e.g. `dag = taskflow()`).
```

> - When you define a task, the `task_id` will default to the name of the function you decorated. If you want to change that, you can simply pass a `task_id` to the decorator as we do in the `extract` task above.
> - Similarly, other task level parameters such as retries or pools can be defined within the decorator (see example with `retries` above).

Reviewer 2: ```suggestion
- When you define a task, the `task_id` will default to the name of the function you decorated. If you want to change that, you can simply pass a `task_id` to the decorator as we do in the `extract` task above. Similarly, other task level parameters such as retries or pools can be defined within the decorator (see the example with `retries` above).
```

- You can decorate a function that is imported from another file with something like the following:

    ```python
    from include.my_file import my_function

    @task
    def taskflow_func():
        my_function()
    ```
    

>     This is recommended in cases where you have lengthy Python functions since it will keep your DAG file easier to read. 

Reviewer 2: ```suggestion
    This is recommended in cases where you have lengthy Python functions since it will make your DAG file easier to read. 
```

### Mixing Decorators with Traditional Operators

> If you have a DAG that uses `PythonOperator` and other operators that don't have decorators, you can easily combine decorated functions and traditional operators in the same DAG. For example, we can add an `EmailOperator` to our DAG above like this:

Reviewer 2: ```suggestion
If you have a DAG that uses `PythonOperator` and other operators that don't have decorators, you can easily combine decorated functions and traditional operators in the same DAG. For example, we can add an `EmailOperator` to the previous example by updating our code to the following:
```

default_args = {
    'start_date': days_ago(1),
}

@dag(schedule_interval='@daily', default_args=default_args, catchup=False)
def taskflow():

    @task(task_id='extract', retries=2)
    def extract_bitcoin_price() -> Dict[str, float]:
        return requests.get(API).json()['bitcoin']

    @task(multiple_outputs=True)
    def process_data(response: Dict[str, float]) -> Dict[str, float]:
        logging.info(response)
        return {'usd': response['usd'], 'change': response['usd_24h_change']}

    @task
    def store_data(data: Dict[str, float]):
        logging.info(f"Store: {data['usd']} with change {data['change']}")

    email_notification = EmailOperator(
        task_id='email_notification',
        to='noreply@astronomer.io',
        subject='dag completed',
        html_content='the dag has finished'
    )

    store_data(process_data(extract_bitcoin_price())) >> email_notification

dag = taskflow()
```

Note that when adding traditional operators, dependencies are still defined using bitshift operators.

## Astro Project Decorators

> The [`astro` library](https://github.com/astro-projects/astro) provides decorators and modules that allow data engineers to think in terms of data transformations rather than Airflow concepts when writing DAGs. The goal is to allow DAG writers to focus on defining *execution* logic, without having to worry about orchestration logic.

Reviewer 2: ```suggestion
The [`astro` library](https://github.com/astro-projects/astro) provides decorators and modules that allow data engineers to think in terms of data transformations rather than Airflow concepts when writing DAGs. The goal is to allow DAG writers to focus on defining *execution* logic without having to worry about orchestration logic.
```

> The library contains SQL and dataframe decorators that greatly simplify your DAG code and allow you to directly define tasks without boilerplate operator code. It also allows you to transition seamlessly between SQL and Python for transformations, without having to explicitly pass data between tasks or convert the results of queries to dataframes and vice versa. For a full description of functionality, check out the [repo Readme](https://github.com/astro-projects/astro).

Reviewer 2: ```suggestion
The library contains SQL and dataframe decorators that greatly simplify your DAG code and allow you to directly define tasks without boilerplate operator code. It also allows you to transition seamlessly between SQL and Python for transformations without having to explicitly pass data between tasks or convert the results of queries to dataframes and vice versa. For a full description of functionality, check out the [repo Readme](https://github.com/astro-projects/astro).
```

> To use the `astro` library, you need to install the `astro-projects` package in your Airflow environment, and enable pickling (`AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True`). 

Reviewer 2: ```suggestion
To use the `astro` library, you need to install the `astro-projects` package in your Airflow environment and enable pickling (`AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True`). 
```

> To show `astro` in action, we'll take a simple ETL example. We have data on adoptions from two different animal shelters that we need to aggregate, clean, transform, and append to a reporting table. Some of these tasks are better suited to SQL, and some to Python, but we can easily combine both using `astro` functions. The DAG looks like this:

Reviewer 2: ```suggestion
To show `astro` in action, we'll use a simple ETL example. We have data on adoptions from two different animal shelters that we need to aggregate, clean, transform, and append to a reporting table. Some of these tasks are better suited to SQL, and some to Python, but we can easily combine both using `astro` functions. The DAG looks like this:
```

    max_active_runs=1,
    schedule_interval='@daily', 
    default_args={
        'email_on_failure': False,
        'retries': 0,
        'retry_delay': timedelta(minutes=5)
    },
    catchup=False
    )
def animal_adoptions_etl():
    # Define task dependencies
    combined_data = combine_data(center_1=Table('ADOPTION_CENTER_1', conn_id="snowflake", schema='SANDBOX_KENTEND'),
                                        center_2=Table('ADOPTION_CENTER_2', conn_id="snowflake", schema='SANDBOX_KENTEND'))

    cleaned_data = clean_data(combined_data)
    aggregated_data = aggregate_data(cleaned_data,
                                    output_table=Table('aggregated_adoptions', conn_id='snowflake'))
    # Append transformed data to reporting table
    reporting_data = aql.append(
        conn_id="snowflake",
        append_table="aggregated_adoptions",
        columns=["DATE", "CAT", "DOG"],
        main_table="SANDBOX_KENTEND.adoption_reporting",
    )

    aggregated_data >> reporting_data

animal_adoptions_etl_dag = animal_adoptions_etl()
```

![Astro ETL](https://assets2.astronomer.io/main/guides/decorators/astro_etl_graph.png)

The general steps in the DAG are:

> - Start by combining data from our two source tables. We use a `transform()` function since we are running a SQL statement on tables that already exist in our database. We define the source tables with the `Table()` parameter when we call the function (e.g. `center_1=Table('ADOPTION_CENTER_1', conn_id="snowflake", schema='SANDBOX_KENTEND')`).
> - Run another `transform()` function to clean the data; we don't report on guinea pig adoptions in this example, so we remove them from the dataset. Note that each `transform` function will store results in a table in your database. You can specify an `output_table` to store the results in a specific table, or you can let `astro` create a table in a default temporary schema for you as we do in this task by not defining any output.
> - Transform the data by pivoting using Python. Pivoting is notoriously difficult in Snowflake, so we seamlessly switch to Pandas. In this task we do specify an `output_table` that we want the results stored in.
> - Append the results to an existing reporting table using the `append` function. 

Reviewer 2: ```suggestion
1. Combine data from our two source tables. We use a `transform()` function since we are running a SQL statement on tables that already exist in our database. We define the source tables with the `Table()` parameter when we call the function (`center_1=Table('ADOPTION_CENTER_1', conn_id="snowflake", schema='SANDBOX_KENTEND')`).
2. Run another `transform()` function to clean the data; we don't report on guinea pig adoptions in this example, so we remove them from the dataset. Note that each `transform` function will store results in a table in your database. You can specify an `output_table` to store the results in a specific table, or you can let `astro` create a table in a default temporary schema for you by not defining any output.
3. Transform the data by pivoting using Python. Pivoting is notoriously difficult in Snowflake, so we seamlessly switch to Pandas. In this task we specify an `output_table` that we want the results stored in.
4. Append the results to an existing reporting table using the `append` function. 
```

> Note that by simply defining our task dependencies when calling the functions (e.g. `cleaned_data = clean_data(combined_data)`), `astro` takes care of passing all context and metadata between the tasks. The result is a DAG where we accomplished some tricky transformations without having to write a lot of Airflow code or messily transition between SQL and Python.

Reviewer 2: ```suggestion
Note that by simply defining our task dependencies when calling the functions (e.g. `cleaned_data = clean_data(combined_data)`), `astro` takes care of passing all context and metadata between the tasks. The result is a DAG where we accomplished some tricky transformations without having to write a lot of Airflow code or transition between SQL and Python.
```

## List of Available Airflow Decorators

> There are a limited number of decorators available to use with Airflow, although more will be added in the future. This list provides a reference of what is currently available so you don't have to dig through source code.

Reviewer 2: ```suggestion
There are a limited number of decorators available to use with Airflow, although more will be added in the future. This list provides a reference of what is currently available so you don't have to dig through source code:
```
